Receiver flag added.

JackPeisel opened the pull request:
Update to config.py to allow for extra flag, update to server.py to update logic for receiver. If receiver flag is at its default value then execute the original emailing logic, altered slightly to reduce code reduplication with the new logic.

Commit:
Receiver flag added.

Update to config.py to allow for extra flag, update to server.py to update logic for receiver

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -153,17 +153,27 @@ def send_emails(self):
+                # We have a receiver set and do not need to check for a user's
+                # email
+                message = message.format(user=receiver, server=server.name,
+                                         age=self.__age, skip=server.name)
```
Rather than duplicate Lines 167-171 at 175-178, let's factor out the message formatting and SMTP code to avoid that repetition.

Also, an aside. By putting all of this in a single loop, does it mean that the same email will be sent N times for N servers all registered to the same user? If so, can we capture a ticket to improve this to only send 1 email per recipient with a list of servers (doesn't need to be in this PR, but it could be super annoying to get 25 emails, one per server).

Commit:
Factored out some email logic, reduced # of emails sent when receiver is specified

JackPeisel commented:
Factored out the message formatting and SMTP code. You're right about the loop, I hadn't thought of that piece and it's a great point. If one person is receiving all the emails then receiving one email for everything is definitely better. I also changed some of the looping to make this happen.

Edit: I will also at a later point update the readme.

Commit:
Fix to emailing when receiver is not set. Message had not been formatting correctly, resulting in duplicate emails.

greg-hellings commented on cloud_cleaner/config.py:
```diff
@@ -24,7 +24,10 @@
-             Only used if --email is set. Required if --email is set'''
+             Only used if --email is set. Required if --email is set''',
+    "receiver": '''The email address that should be used to receive emails.
+             Only used if --email is set. Required if --email is set.'''
```
Is it required? I thought it was only used as a default when a device didn't have an associated user email. If it is required, then we should check that it is, in fact, provided when the `--email` flag is set.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -150,22 +150,49 @@ def send_emails(self):
+
+        for i in range(len(servers)):
+            # Set the values which will send this email(s)
+            if receiver != "":
```
So if I specify a list to send the report to, it will only receive one message. But if I do not and user "Alice" has N > 1 server, she will still receive N emails.

Why don't you change this to an associate array. Then the above loop can just grab the recipient email (from either the `receiver` value or the `user.email` value), and append the value to a list. Then this can just iterate the keys of the array and send out one email to each address.

Something like this, starting at line 155 above:
```python
servers = {}
for server in self.__targets:
    user = conn.get_user_by_id(server.user_id, False)
    recipient = receiver if receiver != "" else user.email
    if recipient in servers:
        servers[recipient]["servers"].append(server.name)
    else:
        servers[recipient] = {"servers": [server.name], "name": user.name}

for email in servers:
    server_names = "\n".join(servers[email]["servers"])
    user = servers[email]["name"]
```
    etc

Commit:
Fix to emailing from feedback

JackPeisel commented:
For config.py, you're right. Laziness on my part from copying format from other args without fully going over it. 

For server.py, That's a great point, I was only thinking of the case where receiver is set, but this also helps in the case where it is not set(and is higher quality code/a better approach regardless). I mostly echoed what you had started, making a few adjustments. Additionally, I made some changes to how the string is formatted to make the received email easier to read.

Commit:
Change in case of launched_at being NONE

A crash would occur at 140(in this version) if launched_at is None, so, for the time being, instead leave this server as it is, so as not make a decision for the user. This could be updated at a later point, by emailing such a server in the same way with the same message as other servers for example.
